File: probe/src/nexus/a2a/mapping.ts
/**
 * A2A task artifact → NexusAnswerResult 매핑 (Phase 1, SPEC §5.4)
 *
 * Phase 0 서버 매핑의 클라이언트 측 역연산. grounding 필드는 무손실,
 * 근거 없음(failed task 또는 evidence 0건)은 null로 강등한다 — "근거 없는 답변 금지"
 * 불변식을 소비자 측에서 그대로 지키고, Nexus 선택성(실패 시 fallback)을 보존한다.
 */

import type { NexusAnswerResult } from '../types.js';
import type { A2AGroundingData, A2APart, A2ATask } from './types.js';
import { TASK_STATE_FAILED } from './types.js';

/** parts에서 첫 text part의 텍스트를 반환한다. */
function textOf(parts: A2APart[]): string {
  const part = parts.find((p) => p.kind === 'text' && typeof p.text === 'string');
  return part?.text ?? '';
}

/** parts에서 첫 data part의 페이로드를 반환한다. */
function dataOf(parts: A2APart[]): A2AGroundingData | null {
  const part = parts.find((p) => p.kind === 'data' && p.data !== undefined);
  return (part?.data as A2AGroundingData | undefined) ?? null;
}

/**
 * A2A Task를 NexusAnswerResult로 매핑한다. 근거가 없으면 null.
 *
 * @param task message/send 결과 Task
 * @returns grounded answer, 또는 null(failed/근거없음/형식오류)
 */
export function mapTaskToAnswerResult(task: A2ATask | null | undefined): NexusAnswerResult | null {
  if (!task || task.status?.state === TASK_STATE_FAILED) {
    return null;
  }

  const artifact = task.artifacts?.[0];
  if (!artifact || !Array.isArray(artifact.parts)) {
    return null;
  }

  const data = dataOf(artifact.parts);
  const evidence = Array.isArray(data?.evidence) ? data!.evidence : [];
  // 근거 없는 confident answer 금지(Phase 0 §6.1) — 소비자 측 가드.
  if (evidence.length === 0) {
    return null;
  }

  return {
    answer: textOf(artifact.parts),
    evidence_snippets: evidence,
    // Phase 0 DataPart는 graph_findings를 포함하지 않는다 (SPEC §5.4 decision b).
    graph_findings: null,
    provenance: data?.provenance ?? [],
    route_used: typeof data?.route === 'string' ? data.route : '',
    timing_ms: {},
  };
}

File: nexus/nexus/a2a/mapping.py
"""The grounding bridge: ``AnswerResult`` -> A2A task artifact (SPEC §5.4).

This is the heart of the spike. The mapping is **total** and **lossless** for the
grounding fields. A response with no evidence maps to the A2A ``failed`` state — never to
a confident answer without evidence (invariant §6.1). "System decides, LLM narrates":
``confidence`` is a deterministic coarse band derived here, not asserted by the LLM.

Shapes are bound to the pinned A2A SDK (``a2a-sdk==1.1.0``); the JSON-RPC wire form of the
spec types lives in ``a2a.compat.v0_3.types``. The contract tests, not these field names,
are the spec (SPEC §9).
"""

from __future__ import annotations

import uuid

from a2a.compat.v0_3.types import Artifact, DataPart, TaskState, TextPart

from nexus.llm.answer import AnswerResult

# Plain-text reason attached to a failed (no-evidence) task.
NO_EVIDENCE_REASON = "답변을 생성할 수 없습니다 — 근거 부족"


def derive_confidence(evidence_count: int, llm_failed: bool) -> str:
    """A deterministic, coarse confidence band — system-decided, never LLM-asserted.

    More independent corroborating sources ⇒ higher confidence. An LLM failure floors the
    band to ``LOW`` (the evidence is real, but the narration is not). Resolves the SPEC §10
    open question in favour of a coarse band over a raw score (whose scale is route-specific
    and not meaningfully comparable across BM25/vector/graph).
    """
    if evidence_count <= 0:
        return "LOW"
    if llm_failed:
        return "LOW"
    if evidence_count >= 3:
        return "HIGH"
    if evidence_count == 2:
        return "MEDIUM"
    return "LOW"


def _evidence_payload(result: AnswerResult) -> list[dict]:
    """Map ``evidence_snippets`` -> the data part ``.evidence[]`` shape (SPEC §5.4)."""
    return [
        {
            "doc_title": s.get("doc_title", ""),
            "section_path": s.get("section_path", ""),
            "text": s.get("text", ""),
            "score": s.get("score"),
            "source_uri": s.get("source_uri", ""),
        }
        for s in result.evidence_snippets
    ]


def build_grounded_artifact(
    result: AnswerResult,
    tenant: str,
    clearance: str,
) -> tuple[dict, str, str | None]:
    """Map a Nexus grounded answer to an A2A artifact + task state.

    Returns ``(artifact_json, task_state, reason)`` where ``artifact_json`` is a
    spec-shaped (camelCase) dict that round-trips through the SDK ``Artifact`` model,
    ``task_state`` is the A2A ``TaskState`` wire value, and ``reason`` is a plain-text
    failure reason (only when the task failed, else ``None``).

    - **No evidence ⇒ ``failed``** with ``NO_EVIDENCE_REASON``; a confident answer is never
      emitted without ≥1 evidence snippet. The (empty) evidence shape is still attached.
    - **Evidence present ⇒ ``completed``**, even if the LLM narration failed (the answer
      text then carries the fallback + raw evidence, and confidence is floored to LOW).
    """
    evidence = _evidence_payload(result)
    has_evidence = len(evidence) > 0

    data = {
        "evidence": evidence,
        "provenance": [
            {
                "source_uri": p.get("source_uri", ""),
                "source_version": p.get("source_version", ""),
                "doc_rid": p.get("doc_rid", ""),
                # SPEC §5.4: the accountable-review stamp, so a consumer (Probe's
                # SpecRef.approvedHash) can verify retrieval against Arbiter's hash.
                "approved_hash": p.get("approved_hash", ""),
            }
            for p in result.provenance
        ],
        "confidence": derive_confidence(len(evidence), result.llm_failed),
        "route": result.route_used,
        "policy": {"tenant": tenant, "clearance_applied": clearance},
    }

    artifact = Artifact(
        artifact_id=uuid.uuid4().hex,
        name="grounded_answer",
        parts=[TextPart(text=result.answer), DataPart(data=data)],
    )
    artifact_json = artifact.model_dump(mode="json", by_alias=True, exclude_none=True)

    if has_evidence:
        return artifact_json, TaskState.completed.value, None
    return artifact_json, TaskState.failed.value, NO_EVIDENCE_REASON
